rfd/0015-2fa-management.md

```diff
@@ -0,0 +1,254 @@
+the `u2f-host` dependency.
+
+### UX
+
+For the below examples, we assume that the user already has at least one 2FA
+device enrolled. The [bootstrap](#bootstrap) section covers registration of the
+first device.
+
+#### CLI
+
+Login:
+
+```sh
+$ tsh login --login=non-sso-user
+Enter password for Teleport user non-sso-user: ...
+Tap your security key... <tap>
+
+$ tsh login --login=non-sso-user --mfa=otp
+Enter password for Teleport user non-sso-user: ...
+Enter OTP code: ...
+
+$ tsh login --login=sso-user --auth=oidc
+# SSO page opens
+# no 2FA prompt from Teleport
+```
+
+2FA management:
+
+```sh
+$ tsh mfa ls
+MFA deivice name   Type   ID                                     Added at                          Last used
+----------------   ----   ------------------------------------   -------------------------------   -------------------------------
+android OTP        OTP    fa004bf4-acc7-435d-8965-5f5a0a4552e8   Tue 08 Dec 2020 01:29:42 PM PST   Tue 15 Dec 2020 01:29:42 PM PST
+yubikey            U2F    c8fbb126-3a29-4a9c-bfe9-ebaed31d8585   Wed 09 Dec 2020 02:00:13 PM PST   Wed 16 Dec 2020 02:00:13 PM PST
+
+$ tsh mfa add
+Adding a new MFA device.
+Choose device type (1 - OTP, 2 - U2F): 2
+Enter device name: solokey
+Tap any *registered* security key... <tap>
+Tap your *new* security key... <tap>
+MFA device "solokey" added.
+
+$ tsh mfa ls
+MFA deivice name   Type   ID                                     Added at                          Last used
+----------------   ----   ------------------------------------   -------------------------------   -------------------------------
+android OTP        OTP    fa004bf4-acc7-435d-8965-5f5a0a4552e8   Tue 08 Dec 2020 01:29:42 PM PST   Tue 15 Dec 2020 01:29:42 PM PST
+yubikey            U2F    c8fbb126-3a29-4a9c-bfe9-ebaed31d8585   Wed 09 Dec 2020 02:00:13 PM PST   Wed 16 Dec 2020 02:00:13 PM PST
+solokey            U2F    87d4fb03-012c-451c-ab0d-5f2a681c119a   Wed 16 Dec 2020 02:05:46 PM PST   Wed 16 Dec 2020 02:05:46 PM PST
+
+# remove by name
+$ tsh mfa rm yubikey
+Tap any *registered* security key... <tap>
+MFA device "yubikey" removed.
+
+# remove by ID
+$ tsh mfa rm fa004bf4-acc7-435d-8965-5f5a0a4552e8
+Tap any *registered* security key... <tap>
+MFA device "android OTP" removed.
+
+$ tsh mfa ls
+MFA deivice name   Type   ID                                     Added at                          Last used
+----------------   ----   ------------------------------------   -------------------------------   -------------------------------
+solokey            U2F    87d4fb03-012c-451c-ab0d-5f2a681c119a   Wed 16 Dec 2020 02:05:46 PM PST   Wed 16 Dec 2020 02:05:46 PM PST
+
+# If 2FA is optional:
+$ tsh mfa rm solokey
+You are about to remove the only remaining MFA device.
+This will disable MFA during login.
+Are you sure? (y/N): N
+
+# If 2FA is required:
+$ tsh mfa rm solokey
+Can't remove the only remaining MFA device.
+Please add a replacement MFA device first using "tsh mfa add".
+```
+
```

Comment:
I was looping back on older issues, there is this use case for reseting/regenerating a users OTP without having to re-create the user. https://github.com/gravitational/teleport/issues/2280  It's an edge case, and maybe less important now users can setup more than one token.

Reply:
Yeah, I think registering a new OTP device and removing the old one works.
A user will need to have access to one of the registered OTP or U2F devices though, otherwise there's a risk of 2FA bypass.